File: src/main/java/io/github/shaksternano/mediamanipulator/io/ImageProcessor.java
package io.github.shaksternano.mediamanipulator.io;

import java.awt.image.BufferedImage;
import java.io.Closeable;
import java.io.IOException;

public interface ImageProcessor<T> extends Closeable {

    BufferedImage transformImage(BufferedImage image, FrameData frameData, T globalData) throws IOException;

    T globalData(BufferedImage image) throws IOException;

    default boolean isDone(boolean readAllFrames, FrameData frameData, T globalData) {
        return readAllFrames;
    }

    @Override
    default void close() throws IOException {
    }
}

File: core/src/main/kotlin/io/github/shaksternano/borgar/core/media/ImageProcessor.kt
package io.github.shaksternano.borgar.core.media

import io.github.shaksternano.borgar.core.io.SuspendCloseable
import kotlinx.coroutines.flow.Flow
import java.awt.image.BufferedImage

interface ImageProcessor<T : Any> : SuspendCloseable {

    suspend fun constantData(firstFrame: ImageFrame, imageSource: Flow<ImageFrame>, outputFormat: String): T

    suspend fun transformImage(frame: ImageFrame, constantData: T): BufferedImage

    override suspend fun close() = Unit
}

class SimpleImageProcessor(
    private val transform: (ImageFrame) -> BufferedImage,
) : ImageProcessor<Unit> {

    override suspend fun constantData(firstFrame: ImageFrame, imageSource: Flow<ImageFrame>, outputFormat: String) =
        Unit

    override suspend fun transformImage(frame: ImageFrame, constantData: Unit): BufferedImage = transform(frame)
}
